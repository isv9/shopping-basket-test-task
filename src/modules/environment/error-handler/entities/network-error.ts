export class NetworkError extends Error {
    constructor() {
        super('Network error');
        this.name = 'NetworkError';
    }
}
