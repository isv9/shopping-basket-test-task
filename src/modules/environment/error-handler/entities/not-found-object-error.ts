export class NotFoundObjectError extends Error {
    constructor(objectName: string) {
        super(`Not found '${objectName}'`);
        this.name = 'NotFoundObjectError';
    }
}
