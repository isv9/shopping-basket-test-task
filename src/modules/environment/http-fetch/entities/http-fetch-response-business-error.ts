export type HttpFetchResponseBusinessError = { error: string; type: string };