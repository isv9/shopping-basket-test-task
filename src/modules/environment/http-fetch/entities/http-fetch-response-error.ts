import {HttpFetchResponse} from "./http-fetch-response";

export interface HttpFetchResponseError<TErrorBody = any> extends Error {
    request?: any;
    response: HttpFetchResponse<TErrorBody>;
    code?: string;
}