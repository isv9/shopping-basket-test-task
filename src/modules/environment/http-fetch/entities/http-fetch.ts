import { HttpFetchResponse } from './http-fetch-response';

export interface HttpFetch {
    get<TResponse = object>(url: string): Promise<HttpFetchResponse<TResponse>>;

    post<TBody, TResponse = object>(url: string, data: TBody): Promise<HttpFetchResponse<TResponse>>;

    delete<TBody, TResponse = object>(url: string, data: TBody): Promise<HttpFetchResponse<TResponse>>;
}
