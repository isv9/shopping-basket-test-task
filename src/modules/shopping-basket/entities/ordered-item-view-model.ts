export interface OrderedItemViewModel {
    itemId: string;
    name: string;
    orderedCount: number;
}
