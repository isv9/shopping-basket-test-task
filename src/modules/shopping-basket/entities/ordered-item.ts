export interface OrderedItem {
    itemId:string;
    count: number;
}
