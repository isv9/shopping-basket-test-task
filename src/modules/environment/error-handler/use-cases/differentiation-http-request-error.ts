import {HttpFetchResponseError} from "../../http-fetch/entities/http-fetch-response-error";
import {HttpFetchResponse} from "../../http-fetch/entities/http-fetch-response";
import {NetworkError} from "../entities/network-error";
import {NotFoundObjectError} from "../entities/not-found-object-error";
import {BusinessError} from "../entities/business-error";
import {HttpFetchResponseBusinessError} from "../../http-fetch/entities/http-fetch-response-business-error";
import {HttpRequestStatusError} from "../entities/http-reques-status-error";

export function defineHttpRequestError(error?: HttpFetchResponseError<HttpFetchResponseBusinessError>): Error {
    if (error) {
        const {
            response,
            message
        } = error;
        if (message === 'Network Error') {
            return new NetworkError();
        }
        if (response) {
            return defineHttpRequestErrorByStatus(response);
        }
        return new NotFoundObjectError('response in server error');
    }
    return new NotFoundObjectError('server error');
}

export function defineHttpRequestErrorByStatus(response: HttpFetchResponse<HttpFetchResponseBusinessError>): Error {
    const {
        status,
        statusText
    } = response;
    if (!status) {
        return new NotFoundObjectError('status server error');
    }
    switch (status) {
        case 400: {
            const {
                data
            } = response;
            if (!data) {
                return new NotFoundObjectError('server business error');
            }
            return new BusinessError(data.error);
        }
        // case 408:
        // case 504: {
        //     return new TimeoutServerError(status);
        // }
        // and ...
        default: {
            return new HttpRequestStatusError(status, statusText);
        }
    }
}
