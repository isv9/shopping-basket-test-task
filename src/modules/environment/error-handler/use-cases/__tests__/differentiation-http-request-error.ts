import {defineHttpRequestError, defineHttpRequestErrorByStatus} from "../differentiation-http-request-error";
import {NetworkError} from "../../entities/network-error";
import {HttpFetchResponse} from "../../../http-fetch/entities/http-fetch-response";
import {HttpFetchResponseError} from "../../../http-fetch/entities/http-fetch-response-error";
import {HttpFetchResponseBusinessError} from "../../../http-fetch/entities/http-fetch-response-business-error";

describe('defineHttpRequestError', () => {

    it('with undefined', () => {
        expect(defineHttpRequestError(undefined)).toMatchSnapshot();
    });

    it('NetworkError', () => {
        const error = {
            message: 'Network Error'
        } as Partial<HttpFetchResponseError<HttpFetchResponseBusinessError>> as HttpFetchResponseError<HttpFetchResponseBusinessError>;
        expect(defineHttpRequestError(error)).toBeInstanceOf(NetworkError);
    });

});

describe('defineHttpRequestErrorByStatus', () => {

    it('without status', () => {
        const error = {} as Partial<HttpFetchResponse<HttpFetchResponseBusinessError>> as HttpFetchResponse<HttpFetchResponseBusinessError>;
        expect(defineHttpRequestErrorByStatus(error)).toMatchSnapshot();
    });

    it('with 400 status', () => {
        const error = {
            data: {
                error: 'test error message'
            },
            status: 400
        } as Partial<HttpFetchResponse<HttpFetchResponseBusinessError>> as HttpFetchResponse<HttpFetchResponseBusinessError>;
        expect(defineHttpRequestErrorByStatus(error)).toMatchSnapshot();
    });

});