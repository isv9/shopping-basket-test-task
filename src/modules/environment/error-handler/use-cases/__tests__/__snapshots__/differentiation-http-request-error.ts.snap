// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineHttpRequestError with undefined 1`] = `[NotFoundObjectError: Not found 'server error']`;

exports[`defineHttpRequestErrorByStatus with 400 status 1`] = `[BusinessError: test error message]`;

exports[`defineHttpRequestErrorByStatus without status 1`] = `[NotFoundObjectError: Not found 'status server error']`;
